docs/design.md

```diff
@@ -0,0 +1,194 @@
+# Design of image update automation
+
+There are three parts to the design, that operate independently, and
+in sympathy with one another:
+
+ - the job runner (type UpdateJob)
+ - the image metadata reflector (types ImageRepository, ImagePolicy)
+ - the automation controller (type ImageAutomation)
+
+Some tooling comes alongside these parts:
+
+ - the `tk-image` command-line tool lets you create the resource
+   mentioned above, and can be used as an extension to the gitops
+   toolkit command-line tool.
+ - the `image-update` image can be used with `kpt fn`, GitHub Actions,
+   and `UpdateJob` to update images in resources within a working
+   directory.
+
+## Update job controller
+
+The job controller runs `UpdateJob` resources, each of which specifies
+an update operation on a git repository. Each run checks out the given
+repo at the specified ref, runs the specified image with the
+arguments, then commits and pushes as specified.
+
+### Integration
+
+The automation controller and the command-line tool `tk-image` create
+`UpdateJob` resources for making changes within a git repository.
+
+The jobs created by the automation controller and the command-line
+tool use `image-update` as the image to run on the git repo.
+
+### Design notes
+
+This could be a more specific `UpdateImageJob`, but it would differ
+only in the update being done, so it is a small step from there to a
+general job.
+
+The motivation for making the jobs separate to the automation is that
+you can then do ad-hoc updates by creating an `UpdateJob` from
+command-line tooling. The downside is that it needs another moving
+part, albeit a generally-useful one.
+
+## Image metadata reflector
+
+The image metadata reflector reconciles `ImageRepository` and
```

Comment:
Will the `ImageAutomation` controller create `ImageRepository` objects on-demand based on what's in the git repo, or is it expected that the user checks in `ImageRepository` and `ImagePolicy` objects in Git that are synced into the cluster using rest of gotk?

Reply:
Yes to the first part of the question (this is explained later). But they can arrive in the cluster by whatever means; one way would be to put definitions in git and sync them, yes.